InteractionBase/src/vis/Menu.cpp

```diff
@@ -0,0 +1,245 @@
+ **
+ **
+ ** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES,
+ ** INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE
+ ** DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+ ** SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ ** SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
+ ** WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+ ** OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ **
+ **********************************************************************************************************************/
+#include "Menu.h"
+
+#include "VisualizationBase/src/declarative/DeclarativeItemDef.h"
+#include "VisualizationBase/src/declarative/DynamicGridFormElement.h"
+#include "VisualizationBase/src/declarative/GridLayouter.h"
+#include "VisualizationBase/src/items/Text.h"
+#include "VisualizationBase/src/cursor/TextCursor.h"
+#include "QGraphicsColorizeEffect"
+#include "VisualizationBase/src/items/Item.h"
+#include "nodes/ViewSwitcherEntry.h"
+
+namespace Interaction {
+
+ITEM_COMMON_DEFINITIONS(Menu, "item")
+
+Menu::Menu(QVector<Model::Node*> selectableNodes,
+			Visualization::Item* target, StyleType* style, int nrOfColumns)
+	: Menu(arrange(selectableNodes, nrOfColumns), target, style)
+{
+}
+
+Menu::Menu(QVector<QVector<Model::Node *> > selectableNodes,
+			Visualization::Item *target, StyleType *style)
+	: Super(nullptr, style), target_(target), currentNodes_(selectableNodes)
+{
+	mousePosition_ = target->scene()->lastMouseHoverPosition();
+	selectedEffect_ = new QGraphicsColorizeEffect();
+	if (hasTextField())
+	{
+		textField_ = new Visualization::Text(this, &style->nameField(), "Enter name here");
+		textField_->setEditable(true);
+	}
+}
+
+Menu::~Menu()
+{
+	//Move the cursor to original owner
+	target_->moveCursor();
+	//Shouldn't be destroyed
+	target_ = nullptr;
+	//Destroyed by DeclarativeItem
+	focusedItem_ = nullptr;
+	SAFE_DELETE_ITEM(textField_);
+	//Destroyed when the focused item is destroyed
+	if (!focusedItem() || focusedItem()->graphicsEffect() != selectedEffect_)
+		SAFE_DELETE(selectedEffect_);
+	selectedEffect_ = nullptr;
+}
+
+void Menu::initializeForms()
+{
+	auto nodeGrid = (new Visualization::DynamicGridFormElement())
+			->setSpacing(5, 5)
+			->setMajorAxis(Visualization::GridLayouter::ColumnMajor)
+			->setNodesGetter([](Visualization::Item* v)
+				{ auto self = static_cast<I*>(v);
+				  return self->currentNodes(); });
+	//First form: only the nodes
+	addForm(nodeGrid);
+	//Second form: the nodes and a text field
+	addForm((new Visualization::GridLayoutFormElement())
```

Comment:
So I still, don't like the idea of a separate text box. If a node is selected in the grid, then you could allow the user to press F2 and edit the node's name directly in the grid element (as opposed to a separate box). This way we'll simplify this class a little bit, while at the same time making things more beautiful.